pkg/manager/manager.go

```diff
@@ -0,0 +1,160 @@
+		ExitIfError(err, "unable to create a kubernetes client")
+
+		// create a owner ref on the mutating webhook
+		// this way when we remove the statefulset of the manager the webhook will be also removed from the cluster
+		err = webhook.CreateOwnerRefForMutatingWebhook(k.clientset)
+		ExitIfError(err, "unable to create owner reference for mutating webhook object")
+
+		isKubevirtInstalled := checkForKubevirt(k.clientset)
+		poolManager, err := poolmanager.NewPoolManager(k.clientset, rangeStart, rangeEnd, isKubevirtInstalled)
+		ExitIfError(err, "unable to create pool manager")
+		if !isKubevirtInstalled {
+			log.Info("kubevirt was not found in the cluster start a watching process")
+			go k.waitForKubevirt()
+		}
+		go k.waitForSignal()
+
+		log.Info("Setting up manager")
+		mgr, err := manager.New(k.config, manager.Options{MetricsBindAddress: k.metricsAddr})
+		ExitIfError(err, "unable to set up manager")
+
+		log.Info("Setting up controller")
+		ExitIfError(controller.AddToManager(mgr, poolManager), "unable to register controllers to the manager")
+
+		ExitIfError(webhook.AddToManager(mgr, poolManager), "unable to register webhooks to the manager")
+
+		ExitIfError(mgr.Start(k.restartChannel), "unable to run the manager")
+
+		// restart channels
+		k.restartChannel = make(chan struct{})
+		k.kubevirtInstalledChannel = make(chan struct{})
+	}
+}
+
+func checkForKubevirt(kubeClient *kubernetes.Clientset) bool {
+	result := kubeClient.ExtensionsV1beta1().RESTClient().Get().RequestURI("/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/virtualmachines.kubevirt.io").Do()
+	if result.Error() == nil {
+		return true
+	}
+
+	return false
+}
+
+// Check for Kubevirt CRD to be available
+func (k *KubeMacPoolManager) waitForKubevirt() {
+	for _ = range time.Tick(5 * time.Second) {
+		kubevirtExist := checkForKubevirt(k.clientset)
+		log.V(1).Info("kubevirt exist in the cluster", "kubevirtExist", kubevirtExist)
+		if kubevirtExist {
+			close(k.kubevirtInstalledChannel)
+			break
+		}
+	}
+}
+
+// wait for the any interrupt to stop the manager and clean the webhook.
+func (k *KubeMacPoolManager) waitForSignal() {
+	select {
+	// This channel is a system interrupt this will stop the container
+	case <-k.stopSignalChannel:
+		k.continueToRunManager = false
+	// This interrupt occurred when kubevirt is installed in the cluster and will restart the manager code only
+	// The container will not restart in this scenario
+	case <-k.kubevirtInstalledChannel:
+		log.Info("found kubevirt restarting the manager")
+	}
+
+	close(k.restartChannel)
+}
+
+func LoadMacAddressFromEnvVar(envName string) (net.HardwareAddr, error) {
+	if value, ok := os.LookupEnv(envName); ok {
+		poolRange, err := net.ParseMAC(value)
+		if err != nil {
+			return nil, err
+		}
+
+		return poolRange, nil
+	}
+
+	return nil, fmt.Errorf("Environment variable %s don't exist", envName)
+}
+
+func ExitIfError(err error, message string) {
```

Comment:
can we just return an error and handle system interface (return code) in main (if error is returned)